src/mlmc/quantity_concept.py

```diff
@@ -121,120 +325,451 @@ def mult_op(x, y):
+        :return: Quantity
+        """
+        bool_type = BoolType()
+        new_qtype = copy.deepcopy(self.qtype)
+        new_qtype.replace_scalar(bool_type)
 
-    def set(self, range_begin, size):
-        self.begin = range_begin
-        self._size = size
-        self.end = range_begin + size
+        if isinstance(other, (float, int)):
+            assert isinstance(self.qtype.base_qtype(), ScalarType), "Quantities with base type ScalarType " \
+                                                                    "are the only ones that support comparison "
+            return Quantity(quantity_type=new_qtype, input_quantities=[self], operation=op)
+        elif isinstance(other, Quantity):
+            assert self.qtype.size() == other.qtype.size(), "Both quantities must have same structure"
+            assert isinstance(self.qtype.base_qtype(), ScalarType) and isinstance(other.qtype.base_qtype(), ScalarType), \
+                "Quantities with base type ScalarType are the only ones that support comparison "
+            return Quantity(quantity_type=new_qtype, input_quantities=[self, other], operation=op)
+
+    def __lt__(self, other):
+        def lt_op(x, y=other):
+            return self._process_mask(x, y, operator.lt)
+        return self._mask_quantity(other, lt_op)
+
+    def __le__(self, other):
+        def le_op(x, y=other):
+            return self._process_mask(x, y, operator.le)
+        return self._mask_quantity(other, le_op)
+
+    def __gt__(self, other):
+        def gt_op(x, y=other):
+            return self._process_mask(x, y, operator.gt)
+        return self._mask_quantity(other, gt_op)
+
+    def __ge__(self, other):
+        def ge_op(x, y=other):
+            return self._process_mask(x, y, operator.ge)
+        return self._mask_quantity(other, ge_op)
+
+    def __eq__(self, other):
+        def eq_op(x, y=other):
+            return self._process_mask(x, y, operator.eq)
+        return self._mask_quantity(other, eq_op)
+
+    def __ne__(self, other):
+        def ne_op(x, y=other):
+            return self._process_mask(x, y, operator.ne)
+        return self._mask_quantity(other, ne_op)
+
+    def sampling(self, size):
+        """
+        Random sampling
+        :param size: number of samples
+        :return: np.ndarray
+        """
+        def mask_gen(x, *args):
+            indices = np.random.choice(x.shape[1], size=size)
+            mask = np.zeros(x.shape[1], bool)
+            mask[indices] = True
+            return mask
+        return self._mask_quantity(size, mask_gen)
+
+    def __getitem__(self, key):
+        """
+        Get items from Quantity, quantity type must support brackets access
+        :param key: str, int, tuple
+        :return: Quantity
+        """
+        new_qtype = self.qtype[key]  # New quantity type
+
+        reshape_shape = None
+        # ArrayType might be accessed directly regardless of qtype start and size
+        if isinstance(self.qtype, ArrayType):
```

Comment:
Try to avoid usage of isinstance, use overloaded method instead.

Here we can have key acces implemented in the QType, e.g.:

    return Quantity(quantity_type=new_qtype, input_quantities=[self],
         operation=self.qtype._make_getitem_op(...))